packages/central-server/src/kobo/startSyncWithKoBo.js

```diff
@@ -6,98 +6,128 @@
+  let numberOfSurveyResponsesCreated = 0;
+  for (const koboSyncResponse of koboData) {
+    for (const [surveyCode, responses] of Object.entries(koboSyncResponse)) {
+      const survey = await transactingModels.survey.findOne({ code: surveyCode });
+
+      for (const responseData of responses) {
+        if (responseData.eventDate > newSyncTime) {
+          newSyncTime = responseData.eventDate;
+        }
+        const entity = await transactingModels.entity.findOne({ code: responseData.orgUnit });
+        if (!entity) {
+          await dataServiceSyncGroup.log(
+            `Skipping KoBo sync for record id ${responseData.event}: unknown entity ${responseData.orgUnit}`,
           );
+          continue;
         }
+
+        const surveyResponse = await transactingModels.surveyResponse.create({
+          id: generateId(),
+          survey_id: survey.id,
+          user_id: apiUser.user_account_id,
+          assessor_name: responseData.assessor || 'KoBo Integration',
+          entity_id: entity.id,
+          start_time: responseData.eventDate,
+          end_time: responseData.eventDate,
+          data_time: responseData.eventDate,
+        });
+
+        const questions = await transactingModels.question.find({
+          code: Object.keys(responseData.dataValues),
+        });
+        const questionByCode = keyBy(questions, 'code');
+
+        await transactingModels.answer.createMany(
+          Object.entries(responseData.dataValues).map(([questionCode, answer]) => ({
+            id: generateId(),
+            type: questionByCode[questionCode].type,
+            survey_response_id: surveyResponse.id,
+            question_id: questionByCode[questionCode].id,
+            text: answer,
+          })),
+        );
+
+        numberOfSurveyResponsesCreated += 1;
       }
     }
+  }
+
+  return { newSyncTime, numberOfSurveyResponsesCreated };
+};
+
+export async function syncWithKoBo(models, dataBroker, dataServiceSyncGroupCode) {
+  const dataServiceSyncGroup = await models.dataServiceSyncGroup.findOne({
+    service_type: SERVICE_TYPE,
+    code: dataServiceSyncGroupCode,
   });
 
-  // Update sync cursor
-  await models.syncService.update({ id: syncService.id }, { sync_cursor: newSyncTime });
+  if (!dataServiceSyncGroup) {
+    throw new Error(`No KoBo sync service with the code ${dataServiceSyncGroupCode} exists`);
+  }
+
+  try {
+    // Pull data from KoBo
+    const koboData = await dataBroker.pull(
+      {
+        code: dataServiceSyncGroup.config.syncGroupCode,
+        type: dataBroker.getDataSourceTypes().SYNC_GROUP,
+      },
+      {
+        startSubmissionTime: dataServiceSyncGroup.sync_cursor,
+      },
+    );
+
+    // Create new survey_responses in Tupaia
+    const {
+      newSyncTime,
+      numberOfSurveyResponsesCreated,
+    } = await models.wrapInTransaction(async transactingModels =>
+      writeKoboDataToTupaia(transactingModels, koboData, dataServiceSyncGroup),
```

Comment:
Seems like we want to have updating the sync cursor be inside the tx block, otherwise it might create duplicate records in tupaia if sync_cursor fails to update?

Reply:
Ahh good call! Yup will fix this 👍